bentoml/frameworks/tensorflow.py

```diff
@@ -15,79 +22,69 @@ def _is_path_like(p):
     return isinstance(p, (str, bytes, pathlib.PurePath, os.PathLike))
 
 
-def _transform_input_by_tensorspec(_input, tensorspec):
-    '''
-    transform dtype & shape following tensorspec
-    '''
-    try:
-        import tensorflow as tf
-    except ImportError:
-        raise MissingDependencyException(
-            "Tensorflow package is required to use TfSavedModelArtifact"
-        )
-
-    if _input.dtype != tensorspec.dtype:
-        # may raise TypeError
-        _input = tf.dtypes.cast(_input, tensorspec.dtype)
-    return _input
-
-
 class _TensorflowFunctionWrapper:
     '''
     TensorflowFunctionWrapper
     transform input tensor following function input signature
     '''
 
-    def __init__(self, origin_func, fullargspec):
+    def __init__(self, origin_func, arg_names=None, arg_specs=None, kwarg_specs=None):
         self.origin_func = origin_func
-        self.concrete_func = None
-        self.fullargspec = fullargspec
-        self._args_to_indices = {arg: i for i, arg in enumerate(fullargspec.args)}
+        self.arg_names = arg_names
+        self.arg_specs = arg_specs
+        self.kwarg_specs = {k: v for k, v in zip(arg_names or [], arg_specs or [])}
+        self.kwarg_specs.update(kwarg_specs or {})
 
     def __call__(self, *args, **kwargs):
-        signatures = self.origin_func.input_signature
-        if signatures is None:
+        if self.arg_specs is None and self.kwarg_specs is None:
             return self.origin_func(*args, **kwargs)
 
         for k in kwargs:
-            if k not in self._args_to_indices:
+            if k not in self.kwarg_specs:
                 raise TypeError(f"Function got an unexpected keyword argument {k}")
-        signatures_by_kw = {k: signatures[self._args_to_indices[k]] for k in kwargs}
+
+        arg_keys = {k for k, _ in zip(self.arg_names, args)}
```

Comment:
is this a convention in TensorFlow? any reason not to limit this function to keyword argument only?

Reply:
This a convention in Python lol

Reply:
In some special cases, the restored function accepts args and its `arg_names` is None.

Reply:
I mean, using the order of `arg_names`  to map positional argument of `__call__` back to the argument key

Reply:
This behavior seems very fragile to me.  It's good that you added a validation here, although from user perspective, it seems very hard to figure out the sequence of the arguments.  I might be missing some context here tho, that's why I ask if this is a convention in Tensorflow